NullPointerException with StopFilter and simple_query_string with analyze_wildcard if the query contains only stop word prefixes

**Elasticsearch version**: 2.2.0

**JVM version**: OpenJDK Runtime Environment (IcedTea 2.6.4) (7u95-2.6.4-0ubuntu0.14.04.1)
OpenJDK 64-Bit Server VM (build 24.95-b01, mixed mode)

**OS version**: Ubuntu 14.04.1

**Description of the problem including expected versus actual behavior**:
We had an existing API backed by Solr, which we later switched to an Elasticsearch backend. The search endpoint is called repeatedly by our clients with each character typed into a search box (i.e. queries look like `q=t`, `q=te`, `q=tes`, `q=test`, etc. We know about CompletionSuggester, we just haven't figured out how to work into our solution while maintaining backward compatibility with our existing clients).

Until recently we have been using `query_string` for simple keyword searches, appending a '*' to each term to give us a prefix search, and using `analyze_wildcard` (in spite of the performance implications) to still give us reasonable stemming and so on. Now we've added a new client to the portfolio with a larger and somewhat different user base, and the queries we've been seeing have been less forgiving on our simple query sanitisation, so we want to switch to `simple_query_string`. For the most part this is working well, but we've been seeing NullPointerExceptions when the query string consists purely of stop words.

**Steps to reproduce**:
    #!/bin/bash
    
    export ELASTICSEARCH_ENDPOINT="http://localhost:9200"
    
    # Create indexes
    
    curl -XPUT "$ELASTICSEARCH_ENDPOINT/play" -d '{
        "mappings": {
            "thing": {
                "properties": {
                    "prop": {
                        "type": "string",
                        "analyzer": "stop"
                    }
                }
            }
        }
    }'
    
    # Index documents
    curl -XPOST "$ELASTICSEARCH_ENDPOINT/_bulk?refresh=true" -d '
    {"index":{"_index":"play","_type":"thing"}}
    {"prop":"Some text"}
    '
    
    # Do searches
    
    curl -XPOST "$ELASTICSEARCH_ENDPOINT/_search?pretty" -d '
    {
        "query": {
            "simple_query_string": {
                "query": "the*",
                "fields": [
                    "prop"
                ],
                "analyze_wildcard": true
            }
        }
    }
    '

**Provide logs (if relevant)**:
```
[2016-05-08 23:11:16,115][DEBUG][action.search.type       ] [Aireo] [play][1], node[-bdul1V7QgKIWXKi-HTkgQ], [P], v[2], s[STARTED], a[id=_0myvoG0Q2yfqC6jIFCsgg]: Failed to execute [org.elasticsearch.action.search.SearchRequest@458bdcb7] lastShard [true]
RemoteTransportException[[Aireo][127.0.0.1:9300][indices:data/read/search[phase/query]]]; nested: SearchParseException[failed to parse search source [
{
    "query": {
        "simple_query_string": {
            "query": "the*",
            "fields": [
                "prop"
            ],
            "analyze_wildcard": true
        }
    }
}
]]; nested: NullPointerException;
Caused by: SearchParseException[failed to parse search source [
{
    "query": {
        "simple_query_string": {
            "query": "the*",
            "fields": [
                "prop"
            ],
            "analyze_wildcard": true
        }
    }
}
]]; nested: NullPointerException;
        at org.elasticsearch.search.SearchService.parseSource(SearchService.java:853)
        at org.elasticsearch.search.SearchService.createContext(SearchService.java:652)
        at org.elasticsearch.search.SearchService.createAndPutContext(SearchService.java:618)
        at org.elasticsearch.search.SearchService.executeQueryPhase(SearchService.java:369)
        at org.elasticsearch.search.action.SearchServiceTransportAction$SearchQueryTransportHandler.messageReceived(SearchServiceTransportAction.java:368)
        at org.elasticsearch.search.action.SearchServiceTransportAction$SearchQueryTransportHandler.messageReceived(SearchServiceTransportAction.java:365)
        at org.elasticsearch.transport.TransportService$4.doRun(TransportService.java:350)
        at org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:37)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1145)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:615)
        at java.lang.Thread.run(Thread.java:745)
Caused by: java.lang.NullPointerException
        at org.elasticsearch.index.query.SimpleQueryParser.newPrefixQuery(SimpleQueryParser.java:131)
        at org.apache.lucene.queryparser.simple.SimpleQueryParser.consumeToken(SimpleQueryParser.java:406)
        at org.apache.lucene.queryparser.simple.SimpleQueryParser.parseSubQuery(SimpleQueryParser.java:212)
        at org.apache.lucene.queryparser.simple.SimpleQueryParser.parse(SimpleQueryParser.java:152)
        at org.elasticsearch.index.query.SimpleQueryStringParser.parse(SimpleQueryStringParser.java:212)
        at org.elasticsearch.index.query.QueryParseContext.parseInnerQuery(QueryParseContext.java:256)
        at org.elasticsearch.index.query.IndexQueryParserService.innerParse(IndexQueryParserService.java:303)
        at org.elasticsearch.index.query.IndexQueryParserService.parse(IndexQueryParserService.java:206)
        at org.elasticsearch.index.query.IndexQueryParserService.parse(IndexQueryParserService.java:201)
        at org.elasticsearch.search.query.QueryParseElement.parse(QueryParseElement.java:33)
        at org.elasticsearch.search.SearchService.parseSource(SearchService.java:836)
        ... 10 more

```

--- a/core/src/main/java/org/elasticsearch/index/query/SimpleQueryParser.java
+++ b/core/src/main/java/org/elasticsearch/index/query/SimpleQueryParser.java
@@ -128,8 +128,10 @@ public class SimpleQueryParser extends org.apache.lucene.queryparser.simple.Simp
             try {
                 if (settings.analyzeWildcard()) {
                     Query analyzedQuery = newPossiblyAnalyzedQuery(entry.getKey(), text);
-                    analyzedQuery.setBoost(entry.getValue());
-                    bq.add(analyzedQuery, BooleanClause.Occur.SHOULD);
+                    if (analyzedQuery != null) {
+                        analyzedQuery.setBoost(entry.getValue());
+                        bq.add(analyzedQuery, BooleanClause.Occur.SHOULD);
+                    }
                 } else {
                     PrefixQuery prefix = new PrefixQuery(new Term(entry.getKey(), text));
                     prefix.setBoost(entry.getValue());

--- a/core/src/test/java/org/elasticsearch/search/query/SimpleQueryStringIT.java
+++ b/core/src/test/java/org/elasticsearch/search/query/SimpleQueryStringIT.java
@@ -346,4 +346,31 @@ public class SimpleQueryStringIT extends ESIntegTestCase {
         assertSearchHits(searchResponse, "1");
     }
 
+    @Test
+    public void testEmptySimpleQueryStringWithAnalysis() throws Exception {
+        // https://github.com/elastic/elasticsearch/issues/18202
+        String mapping = XContentFactory.jsonBuilder()
+                .startObject()
+                .startObject("type1")
+                .startObject("properties")
+                .startObject("body")
+                .field("type", "string")
+                .field("analyzer", "stop")
+                .endObject()
+                .endObject()
+                .endObject()
+                .endObject().string();
+
+        CreateIndexRequestBuilder mappingRequest = client().admin().indices()
+                .prepareCreate("test1")
+                .addMapping("type1", mapping);
+        mappingRequest.execute().actionGet();
+        indexRandom(true, client().prepareIndex("test1", "type1", "1").setSource("body", "Some Text"));
+        refresh();
+
+        SearchResponse searchResponse = client().prepareSearch()
+                .setQuery(simpleQueryStringQuery("the*").analyzeWildcard(true).field("body")).get();
+        assertNoFailures(searchResponse);
+        assertHitCount(searchResponse, 0l);
+    }
 }

